crates/primitives/src/lib.rs

```diff
@@ -41,3 +41,14 @@ pub use self::{
         Hash,
     },
 };
+
+/// An error that comes from the smart contracting language, but not necessarily the smart contract
+/// itself nor the underlying execution environment.
+#[non_exhaustive]
+#[repr(u32)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, ::scale::Encode, ::scale::Decode)]
+#[cfg_attr(feature = "std", derive(::scale_info::TypeInfo))]
+pub enum LangError {
+    /// Failed to read execution input for the dispatchable.
+    CouldNotReadInput = 1u32,
+}
```

Comment:
Aren't we missing `UnknownSelector` here? The one that green initially requested this feature for?

Reply:
Kinda. I mentioned this in one of the `TODO` points:

> Refactor dispatch logic to bubble up specific dispatch error (we can maybe do this
later, at the expense of always having an extra error variant in LangError)

So tl;dr, we can't easily pull out `UnknownSelector` atm.

Reply:
> at the expense of always having an extra error variant in LangError

Which extra error are you talking about specifically?

But in general that is fine. We that we can add new errors to `LangError` without breaking ABI as no one is allowed to exhaustively match on it.

Another thing: Didn't we agree to move this enum to the metadata crate so it is unified across all languages? Or is this TODO.

Reply:
> Which extra error are you talking about specifically?

We'd have `LangError::CouldNotReadInput` as a potential error that doesn't get used if we start exposing more specific errors.

> Another thing: Didn't we agree to move this enum to the metadata crate so it is unified across all languages? Or is this TODO.

We re-export it in the metadata crate. See this [conversation](https://github.com/paritytech/ink/pull/1450/files#r1012870921).

Reply:
Ah ha, now I understand why this is tricky because `UnknownSelector` is raised in `decode_dispatch` and then just swallowed into a ::scale::Error`. So a bigger refactoring to access that error required.

```
impl ::scale::Decode for __ink_ConstructorDecoder {
            fn decode<I>(input: &mut I) -> ::core::result::Result<Self, ::scale::Error>
            where
                I: ::scale::Input,
            {
                <Self as ::ink::reflect::DecodeDispatch>::decode_dispatch(input)
                    .map_err(::core::convert::Into::into)
            }
        }
```